toolchain/check/testdata/where_expr/constraints.carbon

```diff
@@ -19,23 +19,101 @@ interface I {
 
 fn Impls(V:! J where .Self impls I);
 
-fn And(W:! I where .Second impls I and .Member = .Second);
+fn And(W:! I where .Self impls J and .Member == ());
+
+// --- fail_todo_equal_constraint.carbon
+
+library "[[@TEST_NAME]]";
+
+interface N {
+  let P:! type;
+}
+
+// CHECK:STDERR: fail_todo_equal_constraint.carbon:[[@LINE+7]]:27: error: cannot implicitly convert from `{}` to `<associated type in N>`
+// CHECK:STDERR: fn Equal(T:! N where .P = {});
+// CHECK:STDERR:                           ^~
+// CHECK:STDERR: fail_todo_equal_constraint.carbon:[[@LINE+4]]:27: note: type `{}` does not implement interface `ImplicitAs`
+// CHECK:STDERR: fn Equal(T:! N where .P = {});
+// CHECK:STDERR:                           ^~
+// CHECK:STDERR:
+fn Equal(T:! N where .P = {});
+
+// --- fail_todo_associated_type_impls.carbon
 
-// --- todo_check_constraints.carbon
+library "[[@TEST_NAME]]";
+
+interface L {}
+interface M {}
+
+interface K {
+  let Associated:! L;
+}
+
+// CHECK:STDERR: fail_todo_associated_type_impls.carbon:[[@LINE+7]]:36: error: cannot implicitly convert from `<associated L in K>` to `type`
+// CHECK:STDERR: fn AssociatedTypeImpls(W:! K where .Associated impls M);
+// CHECK:STDERR:                                    ^~~~~~~~~~~
+// CHECK:STDERR: fail_todo_associated_type_impls.carbon:[[@LINE+4]]:36: note: type `<associated L in K>` does not implement interface `ImplicitAs`
+// CHECK:STDERR: fn AssociatedTypeImpls(W:! K where .Associated impls M);
+// CHECK:STDERR:                                    ^~~~~~~~~~~
+// CHECK:STDERR:
+fn AssociatedTypeImpls(W:! K where .Associated impls M);
+
+// --- fail_todo_check_rewrite_constraints.carbon
 
 library "[[@TEST_NAME]]";
 
 import library "state_constraints";
 
 // TODO: Should fail since `2` can't be converted to the type of `I.Member`,
 // but constraints are not yet type checked.
-fn TypeMismatch(X:! I where .Member = 2);
+// CHECK:STDERR: fail_todo_check_rewrite_constraints.carbon:[[@LINE+7]]:46: error: cannot implicitly convert from `i32` to `<associated type in I>`
+// CHECK:STDERR: fn RewriteTypeMismatch(X:! I where .Member = 2);
+// CHECK:STDERR:                                              ^
+// CHECK:STDERR: fail_todo_check_rewrite_constraints.carbon:[[@LINE+4]]:46: note: type `i32` does not implement interface `ImplicitAs`
+// CHECK:STDERR: fn RewriteTypeMismatch(X:! I where .Member = 2);
+// CHECK:STDERR:                                              ^
+// CHECK:STDERR:
+fn RewriteTypeMismatch(X:! I where .Member = 2);
+
+// --- fail_left_of_impls_non_type.carbon
+
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_left_of_impls_non_type.carbon:[[@LINE+7]]:32: error: cannot implicitly convert from `i32` to `type`
+// CHECK:STDERR: fn NonTypeImpls(U:! type where 7 impls type);
+// CHECK:STDERR:                                ^
+// CHECK:STDERR: fail_left_of_impls_non_type.carbon:[[@LINE+4]]:32: note: type `i32` does not implement interface `ImplicitAs`
```

Comment:
This may be out of scope, but should this say `ImplicitAs(type)`?

Reply:
Yes to both it being out of scope and should say `ImplicitAs(type)`.